### PR TITLE
[#2] Add View Extension

### PR DESCRIPTION
**Pull request by lydiasama**
https://github.com/nimblehq/android-common-ktx/issues/2

## What happened 👀

Kotlin provides the ability to extend a class with new functionality without having to inherit from the class or use design patterns such as Decorator. This is done via special declarations called extensions.

## Insight 📝

Add the useful extensions from the View class for using:

- [x] Visibility, animation, convertPxToSp, convertSpToPx, enable when match with condition, ...
- [ ] Add Unit Test

## Proof Of Work 📹

N/A

### CODE REVIEW

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt [hunk 1]
```diff
@@ -0,0 +1,140 @@
+package co.nimblehq.common.extensions
+
+import android.graphics.*
+import android.view.View
+import androidx.annotation.DimenRes
+
+/**
+ * Make view visible.
+ *
```
Comment (hoangnguyen92dn) [inline common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:9]: Should we remove this redundant line? 😄

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+fun View.enableWhen(predicate: () -> Boolean) {
+    this.isEnabled = predicate()
+}
+
+fun View.convertPxToSp(px: Int): Float {
+    return px / resources.displayMetrics.scaledDensity
+}
+
+fun View.convertSpToPx(sp: Float): Int {
```
[the remark this answers is not in the document]
Comment (lydiasama) [reply common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:118]: @hoangnguyen92dn I'm sorry but I don't find a @Sp to use. 😄 
Screen Shot 2021-11-21 at 9 41 25 PM

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+fun View.setTopPaddingRes(@DimenRes dimenRes: Int) {
+    setPadding(paddingLeft, resources.getDimensionPixelSize(dimenRes), paddingRight, paddingBottom)
+}
+
+fun View.setTopPadding(padding: Int) {
+    setPadding(paddingLeft, padding, paddingRight, paddingBottom)
+}
+
+fun View.enableWhen(predicate: () -> Boolean) {
```
[the remark this answers is not in the document]
Comment (lydiasama) [reply common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:98]: Updated.

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (hoangnguyen92dn) [reply common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:120]: Hmm... So we could ignore it. The parameter name will show the required value type 😄

**Review by hoangnguyen92dn — APPROVED** [review]
The rest LGTM

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+ * The visibility will active after the delay.
+ *
+ * @param visible The condition to make view visible.
+ * @param delay The delay time in millisecond to affect the view.
+ */
+fun View.visibleOrGone(visible: Boolean, delay: Long) {
+    if (delay <= 0L) {
+        visibleOrGone(visible)
+    } else {
+        postDelayed({ visibleOrGone(visible) }, delay)
```
Comment (doannimble) [inline common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:49]: @lydiasama do we need to check if the context of view is still valid?

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ * Return bitmap with specific size.
+ *
+ * @param resultHeight Height of bitmap result.
+ * @param resultWidth Width of bitmap result.
+ *
+ * @return bitmap with given height and width
+ */
+fun View.getBitmap(resultHeight: Int, resultWidth: Int): Bitmap {
+    val bitmap = Bitmap.createBitmap(resultWidth, resultHeight, Bitmap.Config.ARGB_8888)
```
Comment (doannimble) [inline common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:133]: @lydiasama If the user inputs any negative value, our function will cause the crash.

I suggest to do the check first 🙏

File: common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt (hunk 4 above)
Comment (lydiasama) [reply common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt:49]: @doannimble Did you mention something like this?
https://stackoverflow.com/a/64535290

Comment (doannimble) [reply common-ktx/src/main/java/co/nimblehq/common/extensions/ViewExt.kt]: @lydiasama Exactly!

**Comment by hoangnguyen92dn** [comment]
@doannimble Please approve the PR if no more concern from your side 🙏

**Pull request merged**